### PR TITLE
layers: Rewrite ViewportState checks and tests

### PR DESCRIPTION
**Pull request by krOoze**
- deal with #1685
- report correct object type
- improve? msg texts a bit
- add extension structs to `pNext` check
- use correct printf token where possible
- make tests to match  changes + improve coverage (`nullptr` +
`multiViewport`)
- remove viewport state (and other stuff) from tests that do not use it
- merge some tests together that are too related
- remove unnecessary dynamic state from `PSOLineWidthInvalid`)
- correct invalid `VkViewport` in some places
- correct missing/null `pViewport` and `pScissors` in some places
- update and correct the VUID database

### CODE REVIEW

File: layers/vk_validation_error_database.txt [hunk 1]
```diff
@@ -1503,15 +1503,15 @@ VALIDATION_ERROR_14e09005~^~Y~^~Unknown~^~vkCreateViSurfaceNN~^~VUID-VkViSurface
...
-VALIDATION_ERROR_150009a4~^~N~^~Unknown~^~vkCmdSetViewport~^~VUID-VkViewport-minDepth-01234~^~!(VK_EXT_depth_range_unrestricted)~^~The spec valid usage text states 'minDepth must be between 0.0 and 1.0, inclusive' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkViewport-minDepth-01234)~^~
-VALIDATION_ERROR_150009a6~^~N~^~Unknown~^~vkCmdSetViewport~^~VUID-VkViewport-maxDepth-01235~^~!(VK_EXT_depth_range_unrestricted)~^~The spec valid usage text states 'maxDepth must be between 0.0 and 1.0, inclusive' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkViewport-maxDepth-01235)~^~
+VALIDATION_ERROR_15000996~^~Y~^~ViewportAndScissorBoundsChecking~^~vkCreateGraphicsPipelines,vkCmdSetViewport~^~VUID-VkViewport-width-01227~^~core~^~The spec valid usage text states 'width must be greater than 0.0 and less than or equal to VkPhysicalDeviceLimits::maxViewportDimensions[0]' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkViewport-width-01227)~^~
```
Comment (mark-lunarg) [inline layers/vk_validation_error_database.txt]: @krOoze, the fourth value in the database is generated each time we rebuild the DB from the spec, so editing these is unnecessary and could possibly cause problems (not sure about this).  This field is unused right now and is often misleading in that the generator picks up whatever the current section is for the VU which, as you have discovered, is sometimes inaccurate.  We should probably remove it at some point.

Comment (krOoze) [reply layers/vk_validation_error_database.txt]: It is potentially useful info though. It should be possible to generate this accurately from the `vk.xml`.

Comment (krOoze) [reply layers/vk_validation_error_database.txt]: It should be reset back now.

I also flicked `N` the `VkViewport` checks as I deferred implementing them in `vkCreateGraphicsPipelines`.

File: layers/vk_validation_error_database.txt [hunk 2]
```diff
@@ -1503,13 +1503,13 @@ VALIDATION_ERROR_14e09005~^~Y~^~Unknown~^~vkCreateViSurfaceNN~^~VUID-VkViSurface
...
-VALIDATION_ERROR_1500099a~^~Y~^~ViewportBoundsCheckingWithNVHExtensionEnabled~^~vkCmdSetViewport~^~VUID-VkViewport-height-01229~^~(VK_KHR_maintenance1)~^~The spec valid usage text states 'height must be greater than or equal to -VkPhysicalDeviceLimits::maxViewportDimensions[1] and less than or equal to VkPhysicalDeviceLimits::maxViewportDimensions[1]' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VUID-VkViewport-height-01229)~^~
+VALIDATION_ERROR_15000996~^~N~^~ViewportAndScissorBoundsChecking~^~vkCmdSetViewport~^~VUID-VkViewport-width-01227~^~core~^~The spec valid usage text states 'width must be greater than 0.0 and less than or equal to VkPhysicalDeviceLimits::maxViewportDimensions[0]' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkViewport-width-01227)~^~
```
Comment (tobine) [inline layers/vk_validation_error_database.txt]: Doc validator is complaining that the 6 checks below are still in source. Did you change the PR?

    VALIDATION_ERROR_1500099e
    VALIDATION_ERROR_15000998
    VALIDATION_ERROR_150009a0
    VALIDATION_ERROR_150009a2
    VALIDATION_ERROR_15000996
    VALIDATION_ERROR_1500099a

Comment (krOoze) [reply layers/vk_validation_error_database.txt]: As said above, I changed them to N as they are only partially implemented.

What is the purpose of this "doc validator"?
I think it has it bit backwards. I think it is a problem if it says `Y` on something that is not fully implemented, not other way around. That is assuming purpose of this flag is to track progress, and inform app-makers what is and isn't (fully) checked.

Comment (krOoze) [reply layers/vk_validation_error_database.txt]: I could just implement them though, I just run out of free time when I committed (and the PR seemed good enough to me)...

Comment (krOoze) [reply layers/vk_validation_error_database.txt]: OK, reverted this, so it makes script happy.

Comment (tobine) [reply layers/vk_validation_error_database.txt]: The doc validator checks the state of the database against the actual code. It's primary purpose is to make sure that checks that are in the code are reflected as "Y" in the database. Secondarily it flags issues if "Y" checks don't appear in the code, and also makes sure that test names in the database appear in the test file.

All of its checks are a bit naive, as it just scrubs the source for VUID strings and checks against DB. I agree that items should only be "Y" in the DB if they're completely implemented. That means that in order for doc validator to be happy, if you only partially implement a check, then its VUID should not appear in the code (on a non-commented line) until it's fully implemented and DB can be switched to "Y".

**Comment by tobine** [comment]
This LGTM and is passing testing so I'll merge.
In general this is alot of code for a single CL. In the future, if possible, it would be good to split all of the test changes across multiple CLs to simplify review.

**Comment by mark-lunarg** [comment]
Hey @tobine, something in this (we think) last patch set is causing Nvidia Linux to crash when running DynViewportAndScissorMismatch, can you take a look?

**Comment by tobine** [comment]
Yes I'm looking at this. Crashing on pipeline creation. Also crashing on Nvidia PixelC w/ Android OS.

**Comment by krOoze** [comment]
@tobine That's odd, the set up should be very similar to the other tests. Do they crash too?

**Comment by krOoze** [comment]
Note to self:
- The `VkPipelineRasterizationStateCreateInfo` is weird (`lineWidth?!)
- gp_ci is missing the `shader_stages_size`
- `VkDescriptorSetLayout`s should not be necessary for the tests

**Pull request merged**